snappy/encoder.nim

```diff
@@ -0,0 +1,406 @@
+        let bytesBetweenHashLookups = skip shr 5
+        skip += bytesBetweenHashLookups
+        let nextIp = ip.offset(bytesBetweenHashLookups)
+        if nextIp > ipLimit:
+          ip = nextEmit
+          emitRemainder()
+
+        candidate = baseIp.offset(table[hash])
+
+        table[hash] = uint16 baseIp.distance(ip)
+        if uint32(data) == load32(candidate):
+          break
+        data = load32(nextIp)
+        ip = nextIp
+
+      # A 4-byte match has been found. We'll later see if more than 4 bytes
+      # match. But, prior to the match, src[nextEmit:s] are unmatched. Emit
+      # them as literal bytes.
+      emitLiteral(op, nextEmit, uint32 nextEmit.distance(ip), true)
+
+    # Call emitCopy, and then see if another emitCopy could be our next
+    # move. Repeat until we find no match for the input immediately after
+    # what was consumed by the last emitCopy call.
+    #
+    # If we exit this loop normally then we need to call emitLiteral next,
+    # though we don't yet know how big the literal will be. We handle that
+    # by proceeding to the next iteration of the main loop. We also can
+    # exit this loop via goto if we get close to exhausting the input.
+    while true:
+      # Invariant: we have a 4-byte match at ip, and no need to emit any
+      # literal bytes prior to ip.
+      var base = ip
+
+      let
+        matched = findMatchLength(
+          candidate.offset(4), ip.offset(4), ipEnd, data) + 4
+      ip = ip.offset(matched)
+
+      emitCopy(op, uint16(candidate.distance(base)), matched)
+
+      if ip > ipLimit:
+        emitRemainder()
+
+      # We could immediately start working at s now, but to improve
+      # compression we first update the hash table at s-1 and at s. If
+      # another emitCopy is not our next move, also calculate nextHash
+      # at s+1. At least on ARCH=amd64, these three hash calculations
+      # are faster as one load64 call (with some shifts) instead of
+      # three load32 calls.
+      table[hash(load32(ip.offset(-1)), tableMask)] = uint16(baseIp.distance(ip) - 1)
+
+      let hash = hash(uint32 data, tableMask)
+      candidate = baseIp.offset(table[hash])
+      table[hash] = uint16 baseIp.distance(ip)
+
+      if uint32(data) != load32(candidate):
+        break
+    preload = uint32(data shr 8)
+  emitRemainder()
+
+func encodeFrame*(input: openArray[byte], output: var openArray[byte]): int =
+  # Write a single frame of data using either compressed or uncompressed
+  # frames depending on whether we succeed in compressing the input
+  doAssert input.len > 0 and input.len.uint64 <= maxUncompressedFrameDataLen
+  let
+    ilen = uint32 input.len
+
+  doAssert output.len.uint64 >= maxCompressedLen(ilen)
+
+  let
+    # CRC is computed over the uncompressed data and appears first in the
+    # frame, after the header
+    crc = maskedCrc(input)
+  output[4..7] = crc.toBytesLE()
+
+  if output.len >= minNonLiteralBlockSize:
```

Comment:
output.len >= minNonLiteralBlockSize? 
shouldn't it `input.len >= minNonLiteralBlockSize`?
I don't understand how could it pass tests. flaw in the test suite?

Reply:
good catch! the code is ends up being correct by accident because `encodeBlock` checks `input` and emits a literal instead.. adding a test to make sure that it actually works / roundtrips.

One issue with the test suite perhaps is that it doesn't actually check that we compress things - just that it roundtrips - it used to do this but then the C++ implementation changed and the outputs were no longer the same.